### PR TITLE
Spotify API implementation

### PR DESCRIPTION
**Pull request by janeson332**

### CODE REVIEW

File: src/Spotify/SpotifyAPI.cpp [hunk 1]
```diff
@@ -118,25 +133,278 @@ TResult<Playback> SpotifyAPI::getCurrentPlayback(std::string const &accessToken,
   auto response = client->get("/v1/me/player");
 
   nlohmann::json playbackJson;
-
+  if (response.code == cNoContent) {
+    LOG(INFO) << "[SotifyAPI] in getCurrentPlayback, no content received"
```
Comment (skaupper) [inline src/Spotify/SpotifyAPI.cpp]: typo

File: src/Spotify/SpotifyAPI.cpp [hunk 2]
```diff
@@ -118,25 +133,278 @@ TResult<Playback> SpotifyAPI::getCurrentPlayback(std::string const &accessToken,
...
+std::string SpotifyAPI::stringUrlEncode(std::string const &str) {
+  std::map<char, std::string> const replaceMap = {
+      {' ', "%20"}, {'/', "%2F"}, {';', "%3B"}};
+
+  std::string urlEncoded = "";
+  for (auto &elem : str) {
+    if (replaceMap.find(elem) != replaceMap.end()) {
+      urlEncoded.append(replaceMap.at(elem));
+    }
+    urlEncoded.append(1, elem);
```
Comment (skaupper) [inline src/Spotify/SpotifyAPI.cpp]: there should be an else branch, otherwise the replace char isnt replaced at all

Comment (janeson332) [reply src/Spotify/SpotifyAPI.cpp]: ahh im so stupid :sweat_smile:
wonder why it was working ..

File: jukebox_config.ini [hunk 3]
```diff
@@ -6,3 +6,10 @@ minLogLevel=INFO
 
 [RestAPI]
 port=8888
+
+[Spotify]
+port=8889
+clientID=f589b31542ca45a98c076460a021e086
+clientSecret=0a82fbaa5fae43f0904776d6f695a7ad
+redirectUri=http://localhost:8889/spotifyCallback
+scopes=user-read-private user-read-email app-remote-control user-modify-playback-state
```
Comment (skaupper) [inline jukebox_config.ini]: i needed the scope `user-read-playback-state` too. i wonder why it worked for you without it :thinking:

Comment (janeson332) [reply jukebox_config.ini]: i fetched the token always with the developer api .. (it was faster)

**Comment by janeson332** [comment]
would be nice if we can merge this, so i can test the first implementation of the backend

File: src/Spotify/SpotifyAPI.cpp [hunk 4]
```diff
@@ -118,25 +184,319 @@ TResult<Playback> SpotifyAPI::getCurrentPlayback(std::string const &accessToken,
...
+              << "&type=" << cQueryTypeMap.at(type) << "&market=" << market
+              << "&limit=" << limit << "&offset=" << offset;
+
+  client->SetTimeout(cRequestTimeout);
+  auto response = client->get("/v1/search" + queryStream.str());
+  std::cout << "/v1/search" + queryStream.str() << std::endl;
+  //  std::cout<<response.body<<std::endl;
+  nlohmann::json pagingJson;
+  if (response.code == cNoContent) {
+    LOG(INFO) << "[SotifyAPI] in search, no content received" << std::endl;
```
Comment (skaupper) [inline src/Spotify/SpotifyAPI.cpp]: typo

**Comment by janeson332** [comment]
so created the example and did some documentation work ;)
wanna test it? if not, i will merge ;)
btw you really find each typo :sweat_smile:

**Pull request merged**